sql-plugin/src/main/scala/com/nvidia/spark/rapids/SplitColumnarBatchProcessor.scala

```diff
@@ -0,0 +1,333 @@
+   * - The actual creation of GPU ColumnVectors and merging is handled by the
+   *   SplitBatchIteratorWithPartitionData lazily.
+   *
+   * @param batch           Input ColumnarBatch.
+   * @param splitRowData    Array of arrays containing split row data (RowNums and InternalRows).
+   * @param partitionSchema Schema of partition ColumnVectors.
+   * @return An Iterator of combined ColumnarBatches.
+   */
+  private def splitAndCombineBatchWithPartitionData(
+      batch: ColumnarBatch,
+      splitRowData: Array[Array[(Long, InternalRow)]],
+      partitionSchema: StructType): Iterator[ColumnarBatch] = {
+    withResource(batch) { _ =>
+      val splitIndices = calculateSplitIndices(splitRowData)
+      val splitColumnarBatches = splitColumnarBatch(batch, splitIndices)
+      // Combine the split GPU ColumnVectors with partition ColumnVectors.
+      val combinedBatches = splitColumnarBatches.zip(splitRowData).map {
+        case (splitColumnarBatch, splitRowInfo) =>
+          (0 until splitColumnarBatch.numCols).map(splitColumnarBatch.column).foreach {
+            case gpuCol: GpuColumnVector => gpuCol.incRefCount()
+          }
+          val spillableBatch = SpillableColumnarBatch(splitColumnarBatch,
+            SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
+          new ColumnarBatchPartitionMerger(spillableBatch, partitionSchema, splitRowInfo)
+      }
+      new SplitBatchWithPartitionDataIterator(combinedBatches)
+    }
+  }
+
+  /**
+   * Helper method to calculate split indices for dividing ColumnarBatch based on the row counts.
+   *
+   * Example,
+   * Input - Array of arrays having (row counts, InternalRow) pairs:
+   * [
+   *  [ (10, [abc, ab]), (40, [bc, ab]), (63, [abc, bc]) ],
+   *  [ (7, [abc, bc]), (10, [aa, cc]), (11, [ade, fd]) ],
+   *  [ (4, [ac, gbc]), (7, [aa, tt]), (15, [ate, afd]) ]
+   * ]
+   *
+   * Result -
+   * Row counts for each batch: [ 10+40+30, 7+10+11, 4+7+15 ] = [ 80, 28, 26 ]
+   * Split Indices: [ 80, 108 ]
+   * -> A ColumnarBatch with 134 rows will be split in three batches
+   * having [ 80, 28, 26 ] rows in each.
+   *
+   * @param splitRowData Array of arrays having (row counts, InternalRow) pairs for each partition.
+   * @return A sequence of split indices.
+   */
+  private def calculateSplitIndices(splitRowData: Array[Array[(Long, InternalRow)]]): Seq[Int] = {
+    // Calculate the row counts for each batch
+    val rowCountsForEachBatch = splitRowData.map(partitionData =>
+      partitionData.map {
+        case (rowNum, _) => rowNum
+      }.sum.toInt
+    )
+    // Calculate split indices using cumulative sum
+    rowCountsForEachBatch.scanLeft(0)(_ + _).drop(1).dropRight(1)
+  }
+
+  /**
+   * Split a ColumnarBatch into multiple ColumnarBatches based on specified row counts.
+   */
+  private def splitColumnarBatch(input: ColumnarBatch,
+      splitIndices: Seq[Int]): Seq[ColumnarBatch] = {
+    if (splitIndices.isEmpty) {
+      Seq(input)
```

Comment:
When returning an input from a method that is expected to return a new value, need to increment the refcount to make it semantically like a new output.
```suggestion
      Seq(GpuColumnVector.incRefCounts(input))
```